#!/bin/bash
# declare a name for this job
#PBS -N SRR6860519-pileup
# request the queue for this job
#PBS -q cgsd
# request a total of x processors for this job (y nodes and z processors per node)
#PBS -l nodes=1:ppn=16
# request memory
#PBS -l mem=100gb
# specify walltime
#PBS -l walltime=100:00:00
# out log file
#PBS -o SRR6860519-pileup.out
# err log file
#PBS -e SRR6860519-pileup.err

### Pileup scRNA-seq bam file, using cellsnp-lite and vartrix separately,
### with input vcf generated by cellsnp-lite mode 2
### vcf was just used for comparison of time and memory usage, hence was not filtered
vcf=~/projects/csp-bm/result/srr_020321/genotype_mode2/cellsnp-lite-maf0.1/cellSNP.base.vcf.gz

bam=~/projects/csp-bm/result/srr_020321/genotype_mode2/SRR6860519.bam
fa=~/data/cellranger/refdata-cellranger-GRCh38-1.2.0/fasta/genome.fa
ncores=16
out_dir=~/projects/csp-bm/result/srr_020321/pileup_mode1

# path to cellsnp-lite (v1.2.0, packaged by conda)
bin_cellsnp=~/.anaconda3/envs/CSP/bin/cellsnp-lite

# path to vartrix (v1.1.16, downloaded from Github
# https://github.com/10XGenomics/vartrix/releases/download/v1.1.16/vartrix_linux)
bin_vartrix=~/bin/vartrix

# download raw mixed barcode file
# and then extract barcodes for SRR6860519 (which is donor1-rep1 of GSE112013)
raw_barcode=$out_dir/SRR6860519.raw.barcodes.tsv.gz
barcode=$out_dir/SRR6860519.barcodes.tsv
wget ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE112nnn/GSE112013/suppl/GSE112013_Combined_UMI_table.txt.gz -O $raw_barcode
zcat $raw_barcode 2> /dev/null | \
  head -1 | tr '\t' '\n' | \
  grep '^Donor1-.*-1$' | \
  sed 's/^Donor1-//' | \
  sort -u > $barcode
echo "total barcodes for donor1-rep1 is `wc -l $barcode`"  # 1000 cell barcodes

# run cellsnp-lite
/usr/bin/time --verbose $bin_cellsnp -s $bam -O $out_dir/cellsnp-lite \
  -R $vcf -b $barcode -p $ncores --minMAF 0 --minLEN 0 --minMAPQ 20 \
  --minCOUNT 1 --gzip --cellTAG CB --UMItag UB --exclFLAG 772 \
  > $out_dir/cellsnp-lite.out 2> $out_dir/cellsnp-lite.err

# run vartrix
/usr/bin/time --verbose $bin_vartrix --primary-alignments --umi \
  --mapq 20 -b $bam --bam-tag CB -c $barcode --scoring-method coverage \
  --threads $ncores --ref-matrix $out_dir/vartrix.ref.mtx \
  --out-matrix $out_dir/vartrix.alt.mtx --out-variants $out_dir/vartrix.variants.tsv \
  -v $vcf --fasta $fa > $out_dir/vartrix.out 2> $out_dir/vartrix.err

